pyodk/_endpoints/forms.py

```diff
@@ -95,3 +103,47 @@ def get(
             )
             data = response.json()
             return Form(**data)
+
+    def create_version(
+        self,
+        form_id: str,
+        project_id: Optional[int] = None,
+        definition: Optional[str] = None,
+        attachments: Optional[Sequence[str]] = None,
+        version_updater: Optional[Callable[[str], str]] = None,
+    ) -> str:
```

Comment:
There seemed to be a large overlap in use case and functionality, and a general goal to keep the API small, so it seemed OK to combine the proposed variants. Could you please let me know what names you would like these functions to have, and how to describe them in the docstring / readme? Alternatively, I can change the publish call to not set a version when there's a new form definition. In either case I probably should add tests for the attachment-only use case. I think it's pretty useful for it to succeed in creating a new version no matter what combo of optional kwargs is provided.

Reply:
I see two distinct workflows:
1. Update a form definition as it's being manually worked on. I don't think this will be super common but it's a convenience alternative to updating through the frontend. In that case, the `version` is managed in the form definition and the user should be updating it.
1. Script a form attachment update. This is useful for synchronizing with an external data source or updating entities based on form submissions that have come in. In that case the form version generally doesn't matter (the definition isn't changing anyway).

I'd be ok with only adding the second case above for now.

With two methods, I would call them both `update` and use the following descriptions:
- Update the form definition and optionally its attachment(s). The form definition MUST specify a new `version`.
- Update the form's attachment(s). By default, the form version is updated with a timestamp at time of publish. You can also specify a reference to a function that will be applied to the current form's version. It should return the next intended version (by e.g. adding 1 or changing a suffix).

Here are the cases that a single method would need to address:
- form definition with new version, no version_updater: that should work and the version from the def should be used
- form definition with existing version, no version_updater: backend will error and that error should be passed on (should NOT default to timestamp here I don't think because it could mask an error)
- form definition with new version, version_updater: this is weird and is the big thing I was trying to avoid with the two variants. Detecting the case would require parsing the form definition which we don't want to do. So I think our best option is to ignore the version_updater in that case. 
- form definition with existing version, version_updater: I did not intend this to be supported but I think it's ok. We can detect a [`409.6`](https://github.com/getodk/central-backend/blob/467fc0b81b6a48b794194eb724e5d9d4c07971e7/lib/util/problem.js#L160), get the current version, apply `version_updater`, retry with the new version

I find it easier to separate out the two workflows but as long as those cases are handled reasonably and are tested, you can make the final call.

Reply:
Thanks! This extra detail is helpful.

Python doesn't really have multiple dispatch or method overloading like Java or some other languages. Everything is a dict so (if not doing anything special) with duplicate names the last defined method wins. These features can be simulated with helper libraries, but I think it's an unusual pattern for Python. There is single dispatch in the `functools` library, but then the first args would need to be different e.g. `update(def, xid, attach)` and `update(attach, xid)` and I'm not sure if they can have kwargs or different optional args. Similarly there's `typing.overload` but the implementation ends up in one function with a need to check args, i.e. it's mainly an annotation thing.

So if you would like a single method I'd go for the existing catch-all and making sure it's tested against these use cases. Or as separate methods, maybe they could be called `update_definition` and `update_attachments`?

Reply:
Or how about we omit `version_updater` for now. If a form definition is provided, we enforce that it specify a new version and error otherwise. If a form definition is not provided, we always use a timestamp version. That makes the logic simpler for now and it's probably enough for nearly all cases.